### ISSUE TITLE
AzureDevOps artifact feed has reached free tier limit

### ISSUE THREAD

**Issue by jzabroski** [fluentmigrator/fluentmigrator#1425]
**Describe the bug**
https://dev.azure.com/fluentmigrator/fluentmigrator/_releaseProgress?_a=release-environment-logs&releaseId=46&environmentId=164&jobTimelineRecordIdToSelect=46be2a11-6999-5e05-6807-ab62a9ea9c35&selectTaskWithIndex=3#

```
Error: An unexpected error occurred while trying to push the package with VstsNuGetPush.exe. Exit code(1) and error(Error:
Microsoft.VisualStudio.Services.Content.Common.ArtifactBillingException: Artifact cannot be uploaded because max quantity has been exceeded or the payment instrument is invalid. https://aka.ms/artbilling for details.
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.<HandleResponseAsync>d__53.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.<SendAsync>d__51.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.<SendAsync>d__31.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at VstsNuGetPush.PackageAdder.AddPackageToFeed(FeedCoordinates feed, BlobIdentifierWithBlocks blobId, Boolean muted)
   at VstsNuGetPush.VstsNuGetPushBlobStoreOperation.AddPackageToFeed(BlobIdentifierWithBlocks blobId, Boolean muted)
   at VstsNuGetPush.VstsNuGetPushBlobStoreOperation.Execute()
   at VstsNuGetPush.VstsNuGetPushCommand.MainCore(ILogger logger)
   at VstsNuGetPush.VstsNuGetPushCommand.Main())
Packages failed to publish
```

**Comment by jzabroski** [comment]
@eloekset I'm not sure how we delete old packages, when package feeds are supposed to be immutable?

**Comment by jzabroski** [comment]
@fubar-coder This is another argument for dropping Jet, Sap SqlAnywhere, and possibly DB2... smaller package footprint.

**Comment by fubar-coder** [comment]
@jzabroski I concur that dropping support for databases with bundled native code should reduce the package size for the tools considerably

**Comment by jzabroski** [comment]
Or we just upload a one-time package for those binaries to link to the package.

**Comment by fubar-coder** [comment]
That sadly doesn't work IIRC. The console tools (DotNet.Cli and FM.Console) can reference other packages, but those references aren't usable (directly, without jumping through hoops) from the console tool. `nuget install` pulls the dependencies, but it's a parallel installation (under the `packages` directory). There was a similar (IIRC) problem for `dotnet` tools.

**Comment by jzabroski** [comment]
Yes, the Tools typically require OctoPack compliant behavior. 

I think we should only ship OctoPack compliant packages to Nuget.org. or, just don't ship db2 & wait to see who complains?:yum:

The new MSBuild $(Pkg***) variables would perhaps allow for better plug-in behavior for FluentMigrator.MSBuild

**Comment by eloekset** [comment]
I didn't expect that to happen for a public project. Seems like we need to either ask for raised quotas or find some other free NuGet package host for Open Source source projects.

**Comment by fubar-coder** [comment]
MyGet won't work. I used it previously, but they removed the additional storage space even though I only used it for open source projects. What about GitHub itself?

**Comment by eloekset** [comment]
Yes, [GitHub Packages](https://docs.github.com/en/packages/guides/configuring-dotnet-cli-for-use-with-github-packages) looks free for public projects. I can create a Personal Access Token try to push new packages from the release pipeline and then push all we've got from the Azure Artifacts feed if we feel confident with GitHub Packages.

**Comment by eloekset** [comment]
I've configured the pipeline to use GitHub Packages as the preview feed now. Left to confirm that it works. I tried to initiate a new release from the 3.2.15 tag, but Azure Pipelines has been down for 5 hours now, so we'll have to wait until it's up again.

**Comment by jzabroski** [comment]
@eloekset You're awesome.  Thank you for doing that.  I use GitHub Actions for RazorLight btw and the build matrix feature is super awesome.  I also find it easier to add builds to new projects because it is very copy-paste and text-centered vs. ui-centric AzureDevOps.

**Comment by jzabroski** [comment]
> I tried to initiate a new release from the 3.2.15 tag, but Azure Pipelines has been down for 5 hours now, so we'll have to wait until it's up again.

Can we still do that?  I already pushed 3.2.15 to Nuget.org by bypassing the AzureDevOps package feed step.

**Comment by eloekset** [comment]
Yeah, it should be possible to push that version to the new feed. No conflicts there. Azure DevOps is up again now, and the release just started. Not sure if connection to the new feed is configured correctly.

**Comment by eloekset** [comment]
Two things:

1. I've tried pushing packages to FluentMigrator using a PAT for my own user, but this has failed no matter how I configure the service connection. I think we're going to need a PAT for the FluentMigrator organization, but I can't find where to create one, or I don't have access to do so.
2. Are we sure that GitHub Packages has unlimited storage for public repositories? I've got an organization that I use for my own blog, and that organization has a limitation of 500MB, which is less than Azure Artifacts:
![image](https://user-images.githubusercontent.com/791721/109705781-6e420d00-7b98-11eb-9458-ab5085cefa2e.png)

**Comment by jzabroski** [comment]
I think we need to trim the bloat in AzureDevOps then.

I wonder if there is an API where we can quickly see package sizes

**Comment by eloekset** [comment]
We have retention turned off on our Azure Artifacts. These are the default settings:
![image](https://user-images.githubusercontent.com/791721/109709118-78660a80-7b9c-11eb-8dba-bff3eb4f798a.png)

So if I enable this, we'll start loosing old packages that haven't been downloaded the last 30 days.

And we can setup billing through an Azure Subscription. It's not very expensive for a few additional GBs per month: https://azure.microsoft.com/en-us/pricing/details/devops/azure-devops-services/
![image](https://user-images.githubusercontent.com/791721/109709910-53be6280-7b9d-11eb-9c99-896156f026d3.png)

**Comment by eloekset** [comment]
I can run the script that downloads all versions of all packages again, and store them in a few backups. Then turn on the default retention policies. If someone misses old packages, we can always adjust retention settings and upload old versions from the backups again. I'll notify here when it's done.

**Comment by eloekset** [comment]
Ok, I've downloaded all 423 NuGet packages from the Azure Artifacts feed and stored them on my on-prem server and a private cloud backup service. So it should be safe to delete packages from the feed now.

![image](https://user-images.githubusercontent.com/791721/110218142-4ca69580-7eb8-11eb-97bb-bd4bd84888cb.png)

I'll try the default retention settings of the feed, and we'll see what the impact will be:
![image](https://user-images.githubusercontent.com/791721/110218181-86779c00-7eb8-11eb-9cd9-5eb0608efbca.png)

**Comment by eloekset** [comment]
Since I downloaded all versions of all packages to take a backup of them, the retention settings won't delete anything until in 30 days. I've created a separate Azure Subscription and configured billing for our DevOps organization. Now there's no spending limit on Azure Artifacts, so we should be able to deploy pre-releases to the feed again.

I'll watch this in the coming months. If we're able to get down to under 2GBs again, I can remove the billing subscription from the DevOps Organization.

I've modified the release pipeline to use the Azure Artifacts feed called "fluentmigrator" again instead of the GitHubPackages configuration. I can remove that service connection, since it didn't work anyway.

We can close this issue as soon as we've verified that the modifications on the release pipeline has been reverted and it works as intended again. And we should probably reorganize the **FluentMigrator.Console** NuGet package, so that it's possible to install DB provider packages as needed instead of bundling all provider DLLs with every version of that package.
